// react-youtube-autocomplete/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactTypeaheadComponent2 = require('react-typeahead-component2');

var _reactTypeaheadComponent22 = _interopRequireDefault(_reactTypeaheadComponent2);

var _jsonp = require('jsonp');

var _jsonp2 = _interopRequireDefault(_jsonp);

var _OptionsTemplate = require('./OptionsTemplate');

var _OptionsTemplate2 = _interopRequireDefault(_OptionsTemplate);

var _youtubeFinder = require('youtube-finder');

var _youtubeFinder2 = _interopRequireDefault(_youtubeFinder);

var googleAutoSuggestURL = '//suggestqueries.google.com/complete/search?client=youtube&ds=yt&q=';

var YoutubeAutocomplete = (function (_Component) {
  _inherits(YoutubeAutocomplete, _Component);

  function YoutubeAutocomplete(props) {
    _classCallCheck(this, YoutubeAutocomplete);

    _get(Object.getPrototypeOf(YoutubeAutocomplete.prototype), 'constructor', this).call(this, props);
    this.state = {
      inputValue: ''
    };
  }

  _createClass(YoutubeAutocomplete, [{
    key: 'handleChange',
    value: function handleChange(event) {
      var self = this,
          query = event.target.value,
          url = googleAutoSuggestURL + query;

      this.setState({
        inputValue: query
      });

      (0, _jsonp2['default'])(url, function (error, data) {
        if (error) {
          console.log(error);
        } else {
          var searchResults = data[1];
          self.setState({
            options: searchResults
          });
        }
      });
    }
  }, {
    key: 'onClick',
    value: function onClick(event, optionData) {
      var searchTerm = optionData[0];
      this.setState({
        inputValue: searchTerm
      });
    }
  }, {
    key: 'onOptionChange',
    value: function onOptionChange(event, optionData, index) {
      var self = this,
          searchTerm = optionData[0],
          apiKey = this.props.apiKey,
          maxResults = this.props.maxResults ? this.props.maxResults : '50';

      this.setState({
        inputValue: searchTerm
      });
    }
  }, {
    key: 'onDropDownClose',
    value: function onDropDownClose(event) {
      var self = this,
          searchTerm = this.state.inputValue,
          maxResults = this.props.maxResults <= 50 ? this.props.maxResults : '50',
          YoutubeClient = _youtubeFinder2['default'].createClient({ key: this.props.apiKey }),
          params = {
        part: 'id,snippet',
        type: 'video',
        q: searchTerm,
        maxResults: maxResults
      };

      YoutubeClient.search(params, function (error, results) {
        if (error) return console.log(error);
        self.props.callback(results.items);
      });
    }
  }, {
    key: 'render',
    value: function render() {
      // React components using ES6 classes no longer autobind this to non React methods. In your constructor, add:
      // this.onChange = this.onChange.bind(this)
      // this is why you have to do onChange={this.handleChange.bind(this)}
      return _react2['default'].createElement(
        'div',
        null,
        _react2['default'].createElement(_reactTypeaheadComponent22['default'], {
          inputValue: this.state.inputValue,
          placeholder: this.props.placeHolder,
          className: this.props.className,
          onChange: this.handleChange.bind(this),
          optionTemplate: _OptionsTemplate2['default'],
          options: this.state.options,
          onOptionClick: this.onClick.bind(this),
          onOptionChange: this.onOptionChange.bind(this),
          onDropdownClose: this.onDropDownClose.bind(this)
        })
      );
    }
  }]);

  return YoutubeAutocomplete;
})(_react.Component);

exports['default'] = YoutubeAutocomplete;
module.exports = exports['default'];
